### PR TITLE
[Bug] [connector-jdbc] Nullable Column source have null data could be unexpected results.

### PR DESCRIPTION
**Pull request by mosence**
### Purpose of this pull request

Fix. #5559

### Does this PR introduce _any_ user-facing change?

yes

### How was this patch tested?

Test case see the Issue #5559

Result: 
image

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Review by liugddx — COMMENTED** [review]
Please add some test case

**Comment by mosence** [comment]
> Please add some test case

I don't know how to code this test case. Only simulation and results of bugs were conducted.

**Comment by liugddx** [comment]
You need merge dev and run CI test

**Comment by liugddx** [comment]
> You need merge dev and run CI test

image

**Comment by liugddx** [comment]
> > Please add some test case
> 
> I don't know how to code this test case. Only simulation and results of bugs were conducted.

You can test that insert null and ensure that no errors will be reported when executing the task.

**Comment by mosence** [comment]
> > > Please add some test case
> > 
> > 
> > I don't know how to code this test case. Only simulation and results of bugs were conducted.
> 
> You can test that insert null and ensure that no errors will be reported when executing the task.

The problem is that this bug is a data inconsistency issue, and null values do not cause errors.

**Comment by mosence** [comment]
Ci.

**Review by liugddx — DISMISSED** [review]
LGTM. @ic4y PTAL

**Comment by Hisoka-X** [comment]
LGTM, but could you add some test case to avoid regression? Thanks!

**Comment by mosence** [comment]
> LGTM, but could you add some test case to avoid regression? Thanks!

I have not found any existing test cases for data consistency, and currently I do not have time to rewrite one. If there are similar test cases in the future, I can supplement them.

**Comment by Hisoka-X** [comment]
> > LGTM, but could you add some test case to avoid regression? Thanks!
> 
> I have not found any existing test cases for data consistency, and currently I do not have time to rewrite one. If there are similar test cases in the future, I can supplement them.

OK for me. Could you add a issue for this TODO?

**Review by EricJoy2048 — DISMISSED** [review]

**Comment by Hisoka-X** [comment]
After a second consider, I think you can modify exist e2e test case to cover this PR.  For example, use postgres, in https://github.com/apache/seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc, you can add one row with `null` value data, then check it  in sink postgres table after e2e job fininshed.

**Review by Hisoka-X — CHANGES_REQUESTED** [review]
please add test case

**Comment by mosence** [comment]
> please add test case

There is no existing test case that implements `abstract void compareResult() throws SQLException, IOException;` Interface.

I am really not familiar with this testing framework.

**Comment by liugddx** [comment]
> > > LGTM, but could you add some test case to avoid regression? Thanks!
> > 
> > 
> > I have not found any existing test cases for data consistency, and currently I do not have time to rewrite one. If there are similar test cases in the future, I can supplement them.
> 
> OK for me. Could you add a issue for this TODO?

You need to create an issue, which can be used as another task

**Comment by mosence** [comment]
I submit an issue :P
@liugddx

#5623

**Comment by Hisoka-X** [comment]
I think the code of this PR should be fine and run normally, but we really need automated test cases to avoid regressions. Maybe you can refer to the test code that has been implemented in https://github.com/apache/seatunnel/pull/4812 . cc @choucmei

**Comment by ic4y** [comment]
@mosence I have a small question, why not use rs.getObject(resultSetIndex, Integer.class) approach, which is more concise and easier to understand?

**Comment by mosence** [comment]
> @mosence I have a small question, why not use rs.getObject(resultSetIndex, Integer.class) approach, which is more concise and easier to understand?

Because not all driver support this method.

This method note that:
```
@return an instance of {@code type} holding the column value
@throws SQLException if conversion is not supported, type is null or
        another error occurs. The getCause() method of the
exception may provide a more detailed exception, for example, if
a conversion error occurs
@throws SQLFeatureNotSupportedException if the JDBC driver does not support
this method
@since 1.7
```

When driver support jdk1.6, It will be not available. 

And Ex. MySQL implementation:

```
...
              else if (type.equals(Boolean.class) || type.equals(Boolean.TYPE)) {
                return (T) Boolean.valueOf(getBoolean(columnIndex));

            } else if (type.equals(Integer.class) || type.equals(Integer.TYPE)) {
                return (T) Integer.valueOf(getInt(columnIndex));

            } else if (type.equals(Long.class) || type.equals(Long.TYPE)) {
                return (T) Long.valueOf(getLong(columnIndex));

            } else if (type.equals(Float.class) || type.equals(Float.TYPE)) {
                return (T) Float.valueOf(getFloat(columnIndex));

            } else if (type.equals(Double.class) || type.equals(Double.TYPE)) {
                return (T) Double.valueOf(getDouble(columnIndex));

            }
...
```
If will get value 0 for null Number Type Column.

**Comment by choucmei** [comment]
> I think the code of this PR should be fine and run normally, but we really need automated test cases to avoid regressions. Maybe you can refer to the test code that has been implemented in #4812 . cc @choucmei

Yes, these e2e tests all pass locally, note that one of the other e2e's seems to have a driver that does not support getObject.

**Comment by mosence** [comment]
@liugddx @Hisoka-X @ic4y @choucmei 

I modified the test case for MySQL e2e by changing one of the data to a null value for testing, and performed a null value query in the 'compareResult()' method. As shown in my modifications, strangely, I couldn't find the corresponding row. Can someone help me?

PS. I checked the downlink after the test failed and found no row data。

```
2023-10-20T07:27:00.8462436Z ==================== STDERR end   ====================
2023-10-20T07:28:23.5926268Z 2023-10-20 07:28:23,555 ERROR org.apache.seatunnel.e2e.common.junit.TestLoggerExtension - 
2023-10-20T07:28:23.5928162Z --------------------------------------------------------------------------------
2023-10-20T07:28:23.5929955Z Test org.apache.seatunnel.connectors.seatunnel.jdbc.JdbcMysqlIT.testJdbcDb failed with:
2023-10-20T07:28:23.5932776Z org.opentest4j.AssertionFailedError: Config [/jdbc_mysql_source_and_sink_parallel_upper_lower.conf] Null Columns' Data Count ==> expected: <1> but was: <0> ==> 'select * from seatunnel.sink where `c_decimal_unsigned_30` = ?' Result Row: <Not Row>
2023-10-20T07:28:23.6010923Z 	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:152)
2023-10-20T07:28:23.6076307Z 	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
2023-10-20T07:28:23.6086910Z 	at org.apache.seatunnel.connectors.seatunnel.jdbc.JdbcMysqlIT.compareResult(JdbcMysqlIT.java:279)
2023-10-20T07:28:23.6186085Z 	at org.apache.seatunnel.connectors.seatunnel.jdbc.AbstractJdbcIT.testJdbcDb(AbstractJdbcIT.java:293)
2023-10-20T07:28:23.6188774Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2023-10-20T07:28:23.6192347Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2023-10-20T07:28:23.6194825Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-10-20T07:28:23.6196819Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2023-10-20T07:28:23.6201557Z 	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
2023-10-20T07:28:23.6203456Z 	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
2023-10-20T07:28:23.6206097Z 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
2023-10-20T07:28:23.6208733Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
2023-10-20T07:28:23.6210783Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
2023-10-20T07:28:23.6272867Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestTemplateMethod(TimeoutExtension.java:94)
2023-10-20T07:28:23.6335246Z 	at 
```

**Comment by choucmei** [comment]
> @liugddx @Hisoka-X @ic4y @choucmei
> 
> I modified the test case for MySQL e2e by changing one of the data to a null value for testing, and performed a null value query in the 'compareResult()' method. As shown in my modifications, strangely, I couldn't find the corresponding row. Can someone help me?
> 
> PS. I checked the downlink after the test failed and found no row data。
> 
> ```
> 2023-10-20T07:27:00.8462436Z ==================== STDERR end   ====================
> 2023-10-20T07:28:23.5926268Z 2023-10-20 07:28:23,555 ERROR org.apache.seatunnel.e2e.common.junit.TestLoggerExtension - 
> 2023-10-20T07:28:23.5928162Z --------------------------------------------------------------------------------
> 2023-10-20T07:28:23.5929955Z Test org.apache.seatunnel.connectors.seatunnel.jdbc.JdbcMysqlIT.testJdbcDb failed with:
> 2023-10-20T07:28:23.5932776Z org.opentest4j.AssertionFailedError: Config [/jdbc_mysql_source_and_sink_parallel_upper_lower.conf] Null Columns' Data Count ==> expected: <1> but was: <0> ==> 'select * from seatunnel.sink where `c_decimal_unsigned_30` = ?' Result Row: <Not Row>
> 2023-10-20T07:28:23.6010923Z 	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:152)
> 2023-10-20T07:28:23.6076307Z 	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
> 2023-10-20T07:28:23.6086910Z 	at org.apache.seatunnel.connectors.seatunnel.jdbc.JdbcMysqlIT.compareResult(JdbcMysqlIT.java:279)
> 2023-10-20T07:28:23.6186085Z 	at org.apache.seatunnel.connectors.seatunnel.jdbc.AbstractJdbcIT.testJdbcDb(AbstractJdbcIT.java:293)
> 2023-10-20T07:28:23.6188774Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 2023-10-20T07:28:23.6192347Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 2023-10-20T07:28:23.6194825Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 2023-10-20T07:28:23.6196819Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> 2023-10-20T07:28:23.6201557Z 	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
> 2023-10-20T07:28:23.6203456Z 	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
> 2023-10-20T07:28:23.6206097Z 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
> 2023-10-20T07:28:23.6208733Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
> 2023-10-20T07:28:23.6210783Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
> 2023-10-20T07:28:23.6272867Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestTemplateMethod(TimeoutExtension.java:94)
> 2023-10-20T07:28:23.6335246Z 	at 
> ```

Because seatunnel partitioning requires a non-null partition key, you can add the primary column as the partition key. 
like this https://github.com/apache/seatunnel/pull/4812

**Comment by choucmei** [comment]
> @liugddx @Hisoka-X @ic4y @choucmei
> 
> I modified the test case for MySQL e2e by changing one of the data to a null value for testing, and performed a null value query in the 'compareResult()' method. As shown in my modifications, strangely, I couldn't find the corresponding row. Can someone help me?
> 
> PS. I checked the downlink after the test failed and found no row data。
> 
> ```
> 2023-10-20T07:27:00.8462436Z ==================== STDERR end   ====================
> 2023-10-20T07:28:23.5926268Z 2023-10-20 07:28:23,555 ERROR org.apache.seatunnel.e2e.common.junit.TestLoggerExtension - 
> 2023-10-20T07:28:23.5928162Z --------------------------------------------------------------------------------
> 2023-10-20T07:28:23.5929955Z Test org.apache.seatunnel.connectors.seatunnel.jdbc.JdbcMysqlIT.testJdbcDb failed with:
> 2023-10-20T07:28:23.5932776Z org.opentest4j.AssertionFailedError: Config [/jdbc_mysql_source_and_sink_parallel_upper_lower.conf] Null Columns' Data Count ==> expected: <1> but was: <0> ==> 'select * from seatunnel.sink where `c_decimal_unsigned_30` = ?' Result Row: <Not Row>
> 2023-10-20T07:28:23.6010923Z 	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:152)
> 2023-10-20T07:28:23.6076307Z 	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
> 2023-10-20T07:28:23.6086910Z 	at org.apache.seatunnel.connectors.seatunnel.jdbc.JdbcMysqlIT.compareResult(JdbcMysqlIT.java:279)
> 2023-10-20T07:28:23.6186085Z 	at org.apache.seatunnel.connectors.seatunnel.jdbc.AbstractJdbcIT.testJdbcDb(AbstractJdbcIT.java:293)
> 2023-10-20T07:28:23.6188774Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 2023-10-20T07:28:23.6192347Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 2023-10-20T07:28:23.6194825Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 2023-10-20T07:28:23.6196819Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> 2023-10-20T07:28:23.6201557Z 	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
> 2023-10-20T07:28:23.6203456Z 	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
> 2023-10-20T07:28:23.6206097Z 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
> 2023-10-20T07:28:23.6208733Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
> 2023-10-20T07:28:23.6210783Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
> 2023-10-20T07:28:23.6272867Z 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestTemplateMethod(TimeoutExtension.java:94)
> 2023-10-20T07:28:23.6335246Z 	at 
> ```

Because seatunnel partitioning requires a non-null partition key, you can add the primary column as the partition key. 
like this https://github.com/apache/seatunnel/pull/4812

**Comment by mosence** [comment]
@choucmei 
There are different option `partition_column` between `jdbc_mysql_source_and_sink_parallel.conf` and `jdbc_mysql_source_and_sink_parallel_upper_lower.conf` .
I see it 🤪 . 
Let me have a try yet.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java [hunk 1]
```diff
@@ -242,61 +343,116 @@ Pair<String[], List<SeaTunnelRow>> initTestData() {
...
-                                "test".getBytes(),
-                                "test".getBytes(),
-                                "test".getBytes(),
-                                "f".getBytes(),
-                                bigintValue.add(BigDecimal.valueOf(i)),
-                                decimalValue.add(BigDecimal.valueOf(i)),
-                                decimalValue.add(BigDecimal.valueOf(i)),
-                            });
+            SeaTunnelRow row;
+            if (i == 99) {
```
Comment (choucmei) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: Maybe should remove the if else and add null row outside the loop.

Comment (mosence) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: At the beginning, I also placed it outside and kept reporting errors before placing it inside the loop.

Comment (mosence) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-1/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcMysqlIT.java]: I know why this phenomenon happened

I use the 'getJdbcCase()' method to obtain the 'JdbcCase' class

And this method is the factory method.

**Comment by mosence** [comment]
~~I know why this phenomenon happened~~

~~I use the 'getJdbcCase()' method to obtain the 'JdbcCase' class~~

~~And this method is the factory method.~~

**Comment by mosence** [comment]
I guess that `@TestContainers` will be mutil thread for test. so more task running on sametime. 
The result will be all task result. So in `compare` mehtod had unpredictable error occurred.
So the e2e test I will add in Issue #5623. 

Or someone can solve this problem?
image


It test be OK in offline test.

@liugddx @Hisoka-X @ic4y @choucmei

**Comment by Hisoka-X** [comment]
Could you push the test case code? So that we can find the problem with you.

**Comment by mosence** [comment]
> Could you push the test case code? So that we can find the problem with you.

~~Commit [e1ec769](https://github.com/apache/seatunnel/pull/5560/commits/e1ec7698f2c338219cf25ccb55b162cd741b8c92).~~

**Comment by mosence** [comment]
@Hisoka-X 
And my commit auto test.
I found that postgreSQL is extremely prone to data inconsistency issues: 
```
Test org.apache.seatunnel.connectors.seatunnel.jdbc.JdbcPostgresIT.testAutoGenerateSQL failed with:
org.opentest4j.AssertionFailedError: iterable lengths differ, expected: <1000> but was: <659>
```

**Comment by Hisoka-X** [comment]
> I found that postgreSQL is extremely prone to data inconsistency issues:

Already fixed on dev, could you try merge from dev?

**Comment by mosence** [comment]
> > I found that postgreSQL is extremely prone to data inconsistency issues:
> 
> Already fixed on dev, could you try merge from dev?

This is run on yesterday. Now I update my branch.

**Review by mosence — COMMENTED** [review]
Added tests, but did not match expectations, resulting in other issues and inconsistent data

**Comment by mosence** [comment]
> Could you push the test case code? So that we can find the problem with you.
@Hisoka-X 
I add test case code in PR #5734.

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/utils/JdbcUtils.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+    public static byte[] getBytes(ResultSet resultSet, int columnIndex) throws SQLException {
+        if (null == resultSet.getObject(columnIndex)) {
+            return null;
+        }
+        return resultSet.getBytes(columnIndex);
+    }
+
+    /** Support to LOG for debug. */
+    public static void formatResultSet(ResultSet resultSet, Appendable appendable)
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/utils/JdbcUtils.java]: What's for?

Comment (mosence) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/utils/JdbcUtils.java]: In test print the data. I forgot to delete this code.

**Review by Hisoka-X — APPROVED** [review]

**Comment by Hisoka-X** [comment]
Waiting CI pass

**Comment by Hisoka-X** [comment]
BTW, we should make sure #5734 be merged ASAP.

**Comment by mosence** [comment]
> BTW, we should make sure #5734 be merged ASAP.

Yeah, I agree for you.
The data consistency of integrated tools is quite important.

**Review by hailin0 — APPROVED** [review]

**Pull request merged**